### PR TITLE
Add model-based quantifier instantiation

### PR DESCRIPTION
**Pull request by ajreynol**
This is a straightforward reimplementation of Ge/deMoura from CAV 2009.

On SMT-LIB quantified logics where UF/Arrays are present, 30 sec timeout, against enumerative instantiation (d) and finite model finding (fmf):
```
                             Configuration  #unsat      time    #sat      time #solved    #total
                     cvc5-050322-mbo_fmf   52094   13772.2    4125   2864.09     56219     91390
                       cvc5-050322-mbo_d   63688   25021.3    1526   511.571     65214     91390
                       cvc5-050422-mbo_m   61639   23099.1    3324   1195.65     64963     91390

For unsatisfiable :
                       cvc5-050322-mbo_d       X   12245    2117
                     cvc5-050322-mbo_fmf     651       X    1246
                       cvc5-050422-mbo_m      68   10791      X

For satisfiable :
                       cvc5-050322-mbo_d       X       4       2
                     cvc5-050322-mbo_fmf    2603       X    1041
                       cvc5-050422-mbo_m    1800     240       X
```

Summary: it is significantly worse than enumerative instantiation for UNSAT, and significantly worse than FMF for SAT. Still, it adds 240 new SAT benchmarks wrt FMF.

On SMT-LIB quantified logics without UF/Arrays, against CEGQI (ce) and SyQI (sy):
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                 cvc5-050422-mbo_ce_only   11758   6364.34    1754   3593.39     13512     16319
                  cvc5-050422-mbo_m_only   11109   6530.16    2010   3047.27     13119     16319
                 cvc5-050422-mbo_sy_only   11387   7702.08    1386    5119.6     12773     16319
For unsatisfiable : 
                 cvc5-050422-mbo_ce_only       X     747     567
                  cvc5-050422-mbo_m_only      98       X     158
                 cvc5-050422-mbo_sy_only     196     436       X

For satisfiable : 
                 cvc5-050422-mbo_ce_only       X      74     413
                  cvc5-050422-mbo_m_only     330       X     647
                 cvc5-050422-mbo_sy_only      45      23       X
```

Summary: it is state of the art for satisfiable benchmarks, trails CEGQI for unsatisfiable.  

There are further interesting numbers per-logic in the latter set.

### CODE REVIEW

File: src/theory/quantifiers/inst_strategy_mbqi.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Model-based quantifier instantiation
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__QUANTIFIERS__MODEL_ORACLE_H
+#define CVC5__THEORY__QUANTIFIERS__MODEL_ORACLE_H
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.h:19]:
```suggestion
#ifndef CVC5__THEORY__QUANTIFIERS__INST_STRATEGY_MBQI_H
#define CVC5__THEORY__QUANTIFIERS__INST_STRATEGY_MBQI_H
```

File: src/theory/quantifiers/inst_strategy_mbqi.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  std::unordered_set<Node> d_quantChecked;
+  /** Kinds that cannot appear in queries */
+  std::unordered_set<Kind, kind::KindHashFunction> d_nonClosedKinds;
+};
+
+}  // namespace quantifiers
+}  // namespace theory
+}  // namespace cvc5::internal
+
+#endif /* CVC5__THEORY__QUANTIFIERS__MODEL_ORACLE_H */
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.h:101]:
```suggestion
#endif /* CVC5__THEORY__QUANTIFIERS__INST_STRATEGY_MBQI_H */
```

**Comment by ajreynol** [comment]
BV:
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                 cvc5-050422-mbo_ce_only    5003   1213.55     400   784.318      5403      6045
                  cvc5-050422-mbo_m_only    4771   1976.55     573   612.598      5344      6045
                 cvc5-050422-mbo_sy_only    4827   2133.84     289      1151      5116      6045




For unsatisfiable : 
                 cvc5-050422-mbo_ce_only       X     260     238
                  cvc5-050422-mbo_m_only      28       X     100
                 cvc5-050422-mbo_sy_only      62     156       X

For satisfiable : 
                 cvc5-050422-mbo_ce_only       X       9     116
                  cvc5-050422-mbo_m_only     182       X     284
                 cvc5-050422-mbo_sy_only       5       0       X

**Comment by ajreynol** [comment]
FP:
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                 cvc5-050422-mbo_ce_only    1265    3821.9     222   1961.22      1487      2958
                  cvc5-050422-mbo_m_only    1295   4269.95     279   2026.48      1574      2958
                 cvc5-050422-mbo_sy_only    1371   4320.92     167   2134.12      1538      2958




For unsatisfiable : 
                 cvc5-050422-mbo_ce_only       X      17      16
                  cvc5-050422-mbo_m_only      47       X      20
                 cvc5-050422-mbo_sy_only     122      96       X

For satisfiable : 
                 cvc5-050422-mbo_ce_only       X       0      75
                  cvc5-050422-mbo_m_only      57       X     122
                 cvc5-050422-mbo_sy_only      20      10       X

**Comment by ajreynol** [comment]
LIA:
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                 cvc5-050422-mbo_ce_only     377   37.2558     202   127.128       579       607
                  cvc5-050422-mbo_m_only     187   16.7313     194   42.5266       381       607
                 cvc5-050422-mbo_sy_only     254   225.886     186   464.151       440       607




For unsatisfiable : 
                 cvc5-050422-mbo_ce_only       X     194     123
                  cvc5-050422-mbo_m_only       4       X      11
                 cvc5-050422-mbo_sy_only       0      78       X

For satisfiable : 
                 cvc5-050422-mbo_ce_only       X      13      17
                  cvc5-050422-mbo_m_only       5       X      11
                 cvc5-050422-mbo_sy_only       1       3       X

**Comment by ajreynol** [comment]
NIA:
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                 cvc5-050422-mbo_ce_only       7  0.923939       9   1.37009        16        20
                  cvc5-050422-mbo_m_only       8   12.1009      12   6.59029        20        20
                 cvc5-050422-mbo_sy_only       7   2.10981      12   16.0811        19        20




For unsatisfiable : 
                 cvc5-050422-mbo_ce_only       X       0       0
                  cvc5-050422-mbo_m_only       1       X       1
                 cvc5-050422-mbo_sy_only       0       0       X

For satisfiable : 
                 cvc5-050422-mbo_ce_only       X       0       0
                  cvc5-050422-mbo_m_only       3       X       0
                 cvc5-050422-mbo_sy_only       3       0       X

File: src/theory/quantifiers/inst_strategy_mbqi.cpp [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+    Node t,
+    bool toQuery,
+    std::unordered_map<Node, Node>& cmap,
+    std::map<TypeNode, std::unordered_set<Node> >& freshVarType,
+    const std::map<Node, Node>& mvToFreshVar)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  SkolemManager* sm = nm->getSkolemManager();
+  FirstOrderModel* fm = d_treg.getModel();
+  std::unordered_map<Node, Node>::iterator it;
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.cpp:307]:
```suggestion
  std::unordered_map<const Node, Node>::iterator it;
```

Comment (mpreiner) [reply src/theory/quantifiers/inst_strategy_mbqi.cpp:307]: One of the reasons why `auto` is a safer way for iterators ;-)

File: src/theory/quantifiers/inst_strategy_mbqi.cpp [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        if (toQuery)
+        {
+          // return the fresh variable for this term
+          Node k = sm->mkPurifySkolem(cur, "mbk");
+          freshVarType[cur.getType()].insert(k);
+          cmap[cur] = k;
+          continue;
+        }
+        // converting from query, find the variable that it is equal to
+        std::map<Node, Node>::const_iterator itmv = mvToFreshVar.find(cur);
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.cpp:344]:
```suggestion
        std::map<const Node, Node>::const_iterator itmv = mvToFreshVar.find(cur);
```
Ditto.

File: src/theory/quantifiers/inst_strategy_mbqi.cpp [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+      else if (cur.isVar())
+      {
+        if (!cur.getType().isFirstClass())
+        {
+          // can be e.g. tester/constructor/selector
+          cmap[cur] = cur;
+        }
+        else
+        {
+          std::map<Node, Node>::iterator itm = modelValue.find(cur);
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.cpp:364]:
```suggestion
          std::map<const Node, Node>::iterator itm = modelValue.find(cur);
```

File: src/theory/quantifiers/inst_strategy_mbqi.cpp [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+      {
+        Node rv = fm->getValue(r);
+        Assert(rv.getKind() == kind::UNINTERPRETED_SORT_VALUE);
+        convert(rv, true, tmpConvertMap, freshVarType, mvToFreshVar);
+      }
+    }
+  }
+  // constraint: the skolems of the given type are equal to one of the variables
+  // introduced for uninterpreted sorts
+  std::map<TypeNode, std::unordered_set<Node> >::iterator itk;
```
Comment (mpreiner) [inline src/theory/quantifiers/inst_strategy_mbqi.cpp:165]:
```suggestion
  std::map<const TypeNode, std::unordered_set<Node> >::iterator itk;
```

**Review by mpreiner — COMMENTED** [review]
Overall the `process` method is rather hard to read, it might help to factor out the constraint and instantiation generation in separate methods?

**Comment by ajreynol** [comment]
Agreed, I'll split to separate methods.

File: src/theory/quantifiers/inst_strategy_mbqi.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+{
+ public:
+  InstStrategyMbqi(Env& env,
+                   QuantifiersState& qs,
+                   QuantifiersInferenceManager& qim,
+                   QuantifiersRegistry& qr,
+                   TermRegistry& tr);
+  ~InstStrategyMbqi() {}
+  /** reset round */
+  void reset_round(Theory::Effort e) override;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_mbqi.h:51]: Yes, although this is overriding an existing method. I can do a followup PR to modify the quantifiers modules to conform to guidelines.

**Review by mpreiner — APPROVED** [review]

**Pull request merged**